src/runtimes/node/dynamic_imports.js

```diff
@@ -0,0 +1,28 @@
+const { dirname } = require('path')
+
+const findUp = require('find-up')
+const readPackageJson = require('read-package-json-fast')
+
+const getModulesWithDynamicImports = async (warnings) => {
+  const dynamicImportWarnings = warnings
+    // Unfortunately, esbuild doesn't give us any error codes, so this is the
+    // only way we have to single out this case.
+    .filter(({ text }) => text.includes('will not be bundled because the argument is not a string literal'))
+    .map(async ({ location }) => {
+      try {
+        const directory = dirname(location.file)
+        const packageJsonPath = await findUp('package.json', { cwd: directory })
+        const { name } = await readPackageJson(packageJsonPath)
```

Comment:
But `require(packageJsonPath)` would actually load the module, whereas we simply want to extract its name. Or am I misunderstanding your suggestion?

Reply:
No, it would require the `package.json`, not the package's main file.

Reply:
Doh, of course! I misread the code initially.

As for using `require(packageJsonPath)`, it would indeed work. But we've been using `read-package-json-fast` instead, precisely for performance reasons (for example, in the native modules detection plugin for esbuild). I can switch to a simple `require` though, if you prefer that.

As for `read-pkg-up`, I personally don't see the need to be introducing another dependency for this purpose.

Reply:
Checking the code, it looks like `read-package-json-fast` might actually be slower. It [does `fs.readFile()`](https://github.com/npm/read-package-json-fast/blob/dc4d7fa7a8e50cf697594edd90b43697f1e3e29c/index.js#L5) then uses [a module](https://github.com/npm/json-parse-even-better-errors) that re-reimplements `JSON.parse()` with userland code. In my experience, v8 is optimizing `JSON.parse()` tremendously and it's usually much faster than userland implementations. Also, it normalizes the object after reading it, which is not needed here. Finally, I am not quite sure why, but in my experience, `require('./file.json')` is faster than `readFile()` followed by `JSON.parse()`.

Good point though about consistency. :+1:

Reply:
It's used by [the npm CLI](https://github.com/npm/cli/blob/latest/package.json#L100), and my understanding is that it's capable of reading automatically-generated `package.json` files (i.e. the ones found inside `node_modules`, not the user-friendly top-ones found in a package) at the expense of potentially missing certain fields. This is our use case, since we don't care about anything other than the `name` property.

I remember running some benchmarks when we were optimising the native modules plugin, and this package did seem to be slightly faster than `require`, but with a minimal difference that might not even be statistically significant.

Perhaps we can leave it as is on this PR for consistency and then we revisit its usage as a whole?